src/main/java/com/sps/Servlets/NewRecipeServlet.java

```diff
@@ -0,0 +1,138 @@
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.gson.Gson;
+import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.ArrayList;
+import com.google.sps.data.Recipe;
+import com.google.sps.data.Step;
+
+@WebServlet("/new-recipe")
+public class NewRecipeServlet extends HttpServlet {
+
+  private DatastoreService datastore;
+  private final String TAG = "tag";
+  private final String INGREDIENT = "ingredient";
+  private final String STEP = "step";
+
+  @Override
+  public void init() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /*
+   * When a spin-off is created, this GET method gets the original recipe's data.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String keyString = request.getParameter("key");
+    System.out.println(keyString);
+    Entity recipeEntity = null;
+    try {
+      recipeEntity = datastore.get(KeyFactory.stringToKey(keyString));
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      return;
+    }
+    Recipe original = entityToRecipe(recipeEntity);
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJsonUsingGson(original));  
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Collection<String> searchStrings = new HashSet<>();
+    String name = request.getParameter("name");
+    searchStrings.add(name.toUpperCase());
+    String description = request.getParameter("description");
+    EmbeddedEntity tags = getParameters(request, TAG, searchStrings);
+    EmbeddedEntity ingredients = getParameters(request, INGREDIENT, searchStrings);
+    EmbeddedEntity steps = getParameters(request, STEP, null);
+
+    Entity recipe = new Entity("Recipe");
+    recipe.setProperty("name", name);
+    recipe.setProperty("description", description);
+    recipe.setProperty("tags", tags);
+    recipe.setProperty("ingredients", ingredients);
+    recipe.setProperty("steps", steps);
+    recipe.setProperty("search-strings", new ArrayList<String>(searchStrings));
+    datastore.put(recipe);
+
+    response.sendRedirect("/edit-recipe.html");
+  }
+
+  /**
+   * Gets the parameters for fields that have different numbers of parameters from recipe to recipe.
```

Comment:
I'm confused by this Javadoc. Isn't it expected that it would get parameters for recipes regardless of whether the recipe has 2 ingredients or has 20? Basically, I'm confused as to whether this explanation is redundant or if I'm just understanding it wrong (i.e. its for a different use case) - any clarification would be helpful!

Reply:
Edit: I think it's that you're incrementing the parameter name number..? If so, maybe make that clear in the javadoc; without reading the code, the javadoc doesn't really make sense.